vboxtouch/vboxtouch.cpp

```diff
@@ -205,6 +212,18 @@ VirtualboxTouchScreenHandler::VirtualboxTouchScreenHandler(const QString &specif
 {
     setObjectName("Virtualbox Touch Handler");
 
+    // We need to catch the SIGTERM and SIGINT signals, so that we can do a
+    // proper shutdown of Qt and the lipstick, and avoid crashes, hangs and
+    // reboots (e.g. during user switching).
+    // This actions is similar to actions in the qt5-qpa-hwcomposer-plugin
+    // that is not exists in the emulator so we doing it here.
+    struct sigaction new_action;
+    new_action.sa_handler = exit_qt_gracefully;
```

Comment:
Unfortunately this needs something more - check https://doc.qt.io/archives/qt-5.13/unix-signals.html

Reply:
I believe, because we shutting down the app it is unnecessary. This solution is just like in: https://github.com/mer-hybris/qt5-qpa-hwcomposer-plugin/blob/master/hwcomposer/hwcomposer_context.cpp#L83-L88

Reply:
This is the same as thinking acquiring a lock is not needed in a multi-threaded program while it is shutting down (with multiple threads running). They do emphasize the words "can't" in "You can't call Qt functions from Unix signal handlers." for a reason, not just for fun. The handler will be called asynchronously at any time the scheduler decides to interrupt the process.

The way it is handled in hwcomposer leads to random, very hard to reproduce errors (crashes, hangs and reboots as mentioned there) and should be definitely fixed.

Reply:
Threads analogy is pretty clear to me - thanks for enlightening. Then it must be bug in qt5-qpa-hwcomposer-plugin. 
So I updated and tested this using file descriptor. Similar to https://github.com/sailfishos/lipstick/blob/master/src/homeapplication.cpp#L210.